commit 347abf561b29269b90dd5f907ddaa9494ecc7246
Author: Kalesh Singh <kaleshsingh@google.com>
Date:   Fri Oct 11 14:16:27 2024 -0700

    ltp: Introduce pgsize_helpers.h
    
    On Android the x86_64 userspace page-size is emulated on Android studio
    emulators, in order to faciliate the majority of app developers on
    x86_64 host machines being able to test their apps on non-4KiB page size
    devices.
    
    This cause some ltp test cases to fail on the 16KiB x86_64 android
    emulator.
    
    Introduce some helpers to address this that are effectively noops for
    all other targets except x86_64 with non 4KiB base page size. (This
    should only be android studio x86 16KiB emulators).
    
    Bug: 366098040
    Test: m vts_ltp_test_x86_64
    Change-Id: I778ed2698c5bf5382012b95c32231f478616ce44
    Signed-off-by: Kalesh Singh <kaleshsingh@google.com>

diff --git a/include/pgsize_helpers.h b/include/pgsize_helpers.h
new file mode 100644
index 000000000..7e57a7456
--- /dev/null
+++ b/include/pgsize_helpers.h
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+/*
+ * Copyright (c) 2024 Google LLC. All rights reserved.
+ * Author(s): Kalesh Singh <kaleshsingh@google.com>
+ */
+
+#ifndef PGSIZE_HELPER_H
+#define PGSIZE_HELPER_H
+
+#include <unistd.h>
+
+#if defined(__x86_64__)
+/*
+ * Android emulates the userspace page size on some x86_64 emulators.
+ * The kernel page size still remains 4KiB (0x1000).
+ */
+static inline size_t kernel_page_size(void)
+{
+    return 0x1000;
+}
+#else
+static inline size_t kernel_page_size(void)
+{
+	return getpagesize();
+}
+#endif
+
+/*
+ * NOTE: For all cases except Android x86_64 page size emulators,
+ * kernel_page_size == page_size, and the below macros are effectively
+ * no-ops.
+ */
+
+/* The number of kernel pages covered by size */
+static inline size_t nr_kernel_pages(size_t size)
+{
+    return size / kernel_page_size();
+}
+
+/* The number of kernel pages in a @nr_pages userspace pages */
+static size_t nr_pgs_to_nr_kernel_pgs(size_t nr_pages)
+{
+    return nr_pages * nr_kernel_pages(getpagesize());
+}
+
+/* The number of userspace pages in a @nr_pages kernel pages */
+static size_t nr_kernel_pgs_to_nr_pgs(size_t nr_pages)
+{
+    return nr_pages / nr_kernel_pages(getpagesize());
+}
+
+#endif /* PGSIZE_HELPER_H */
